### ISSUE TITLE
Support for int64 Context Maps

### ISSUE THREAD

**Issue by TheSkyentist** [spacetelescope/drizzlepac#1845]
I am attempting to drizzle more than 32 images together. However, context maps appear to only support a maximum of 31 images (see links below) as they are cast to int32 data types. Is it possible to support 64 bit context maps? In addition, I believe these should be unit32 data types so as to reference 32 different images, since currently the image context is not recorded if a uniqid of 32 is used. 
https://github.com/spacetelescope/drizzlepac/blob/f339f3556285259b702dc1a4e35f8c73ae95f0ae/drizzlepac/adrizzle.py#L205
https://github.com/spacetelescope/drizzlepac/blob/f339f3556285259b702dc1a4e35f8c73ae95f0ae/drizzlepac/adrizzle.py#L207
https://github.com/spacetelescope/drizzlepac/blob/f339f3556285259b702dc1a4e35f8c73ae95f0ae/drizzlepac/adrizzle.py#L628
https://github.com/spacetelescope/drizzlepac/blob/f339f3556285259b702dc1a4e35f8c73ae95f0ae/drizzlepac/adrizzle.py#L728

**Comment by stscijgbot-hstdp** [comment]
_This issue is tracked on JIRA as [HLA-1290](https://jira.stsci.edu/browse/HLA-1290)._

**Comment by mcara** [comment]
Why do you need `int64`? Is it for convenience? If so, you can just merge the two planes from the (current, int32) context image(s) into a single int64 array. This will work as long as you have less than 65 images. What do we do for users who need to drizzle 300 images?

The choice of `int32` is based on the fact that FITS standard  (see [Table 5.2](https://archive.stsci.edu/fits/fits_standard/node39.html#s:man)) directly supports only `int16` and `int32` arrays.

Personally, I would actually prefer `int16` as has finer granularity that can result in memory savings.

**Comment by TheSkyentist** [comment]
I was completely unaware of the FITS standard only supporting int32 and int16 datatypes! This actually makes a lot of sense as to why this is the default behaviour. In short, yes it would be nice for convenience, especially for packages that wrap drizzlepac to mosaic scores of images. 

In practice, Astropy's FITS handling can handle additional datatypes, including uint64:
```python
import numpy as np
from astropy.io import fits

x = np.ones((1024,1024),dtype=np.uint64)
hdu = fits.PrimaryHDU(x)
hdu.writeto('test.fits',overwrite=True)
print(fits.getdata('test.fits').dtype) # Prints uint64
```
Even if this isn't 100% compliant with the FITS standard, though I'm no expert on this subject.

Depending on how much effort the implementation requires, I would imagine a default strategy to store the details in an array with a data type of a minimum necessary size, whether it be uint8, uint16, etc. 

I agree that the current implementation will work with 31 (not 32 as a bit is reserved for the sign, similarly int64 will only work for up to 63 images). Perhaps the short term solution would be to print warning messages in the case that a user processes more than the available size? 

While in principle it might be nice to expand compatibility to arbitrary array sizes, it seems that support 64 total images (with uint64) would require the least amount of effort since uint64 is already implemented in numpy (and going above that gets trickier on our normal 64-bit machines). 

In short, if uniq id is outside the range of possible support values, a warning (or error, with a potential override) should be thrown to alert the user of what is happening. And, for short term solutions to support more than 31 images, it may be useful to support uin64 since this is implemented in NumPy and works with Astropy, while going above this would constitute significant effort.

**Comment by mcara** [comment]
> In short, if uniq id is outside the range of possible support values, a warning (or error, with a potential override) should be thrown to alert the user of what is happening.

That's not how it works. context image array is a 3D array with the "z" axis moving "along the planes". Each plane is a `int32` 2D image of the same shape as output data image. When there are more than 32 images, a new plane is created. Two planes now can support up to 64 images. So the context array shape now is (2, 1024, 2048) (as an example) and in general: (Number of Planes, Ny, Nx).

See https://jwst-pipeline.readthedocs.io/en/latest/jwst/resample/main.html#context-image for more details.

I do not see the need to support any other types: you can convert a 3D array to a 2D as long as it fits into your desired data type.

**Comment by mcara** [comment]
Ah, also, `int32` still supports 32 images since the bit used for the sign can be used to denote contribution from the last (input) image. For example, take a look at the table "Eight-bit integers" in https://en.wikipedia.org/wiki/Two%27s_complement. An unsigned `char` (8-bit number) would be interpreted as an integer -1 if 8 input images have all contributed to the output (resampled) image:
```
1111 1111 = -1
1111 1110 = -2
0111 1111 = 127
```

**Comment by TheSkyentist** [comment]
These are all really great points! I appreciate you taking the time to help explain through this, I definitely should get a lot more familiar with the plane system. It's also a good note about the fact that context maps can take on negative values, as I usually check for positive values to see what areas of my full array are filled, I have been neglecting the 32nd bit. Based on these discussions, I see no reason to amend the codebase. Thanks again!

**Comment by stscijgbot-hstdp** [comment]
_Comment by [Steve Goldman](https://jira.stsci.edu/secure/ViewProfile.jspa?name=sgoldman) on [JIRA](https://jira.stsci.edu/browse/HLA-1290?focusedCommentId=746138#comment-746138):_

Issue discussed on [github](https://github.com/spacetelescope/drizzlepac/issues/1845) and determined it was no longer needed.
